addons/im_livechat/static/src/legacy/public_livechat_chatbot.js

```diff
@@ -0,0 +1,874 @@
+        this._addMessage(message, options);
+        if (this._chatWindow.isFolded() || !this._chatWindow.isAtBottom()) {
+            this._livechat.incrementUnreadCounter();
+        }
+
+        if (!options || !options.skipRenderMessages) {
+            this._renderMessages();
+        }
+    },
+
+    /**
+     * This method will be transformed into a 'debounced' version (see init).
+     *
+     * The purpose is to handle steps of type 'free_input_multi', that will let the user type in
+     * multiple lines of text before the bot goes to the next step.
+     *
+     * Every time a 'keydown' is detected into the input, or every time a message is sent, we call
+     * this debounced method, which will give the user about 10 seconds to type more text before
+     * the next step is triggered.
+     *
+     * First we check if the last message was sent by the user, to make sure we always let him type
+     * at least one message before moving on.
+     */
+    _chatbotAwaitUserInput: function () {
+        const lastMessage = this._messages.length !== 0 ? this._messages[this._messages.length - 1] : null;
+        if (lastMessage && lastMessage.getAuthorID() !== this._livechat.getOperatorPID()[0]) {
+            if (!this._chatbotShouldEndScript()) {
+                this._chatbotSetIsTyping();
+                this.nextStepTimeout = setTimeout(
+                    this._chatbotTriggerNextStep.bind(this), this._chatbotMessageDelay);
+            } else {
+                this._chatbotEndScript();
+            }
+        }
+    },
+
+    /**
+     * When the user first interacts with the bot, we want to make sure to actually post the welcome
+     * messages into the conversation.
+     *
+     * Indeed, before that, they are 'virtual' messages that are not tied to mail.messages, see
+     * #_sendWelcomeChatbotMessage() for more information.
+     *
+     * Posting them as real messages allows to have a cleaner model and conversation, that will be
+     * kept intact when changing page on the website.
+     *
+     * It also allows tying any first response / question_selection choice to a chatbot.message
+     * that has a linked mail.message.
+     */
+    _chatbotCheckWelcomeMessages: async function () {
+        const welcomeMessages = this._messages.filter((message) => {
+            return message._id && typeof message._id === 'string' && message._id.startsWith('_welcome_');
+        });
+
+        if (welcomeMessages.length === 0) {
+            // we already posted the welcome messages, nothing to do
+            return;
+        }
+
+        const postedWelcomeMessages = await session.rpc('/im_livechat/chatbot_post_welcome_messages', {
+            channel_uuid: this._livechat.getUUID(),
+            chatbot_script_id: this._chatbot.chatbot_script_id,
+        });
+
+        this._messages = this._messages.filter((message) => {
+            !welcomeMessages.map(welcomeMessage => welcomeMessage._id).includes(message._id);
+        });
```

Comment:
To decrease the complexity (O(n^2) => O(n))

```suggestion
        const welcomeMessagesIds = welcomeMessages.map(welcomeMessage => welcomeMessage._id)
        this._messages = this._messages.filter((message) => {
            !welcomeMessagesIds.includes(message._id);
        });
```